local-cluster/tests/local_cluster.rs

```diff
@@ -1680,7 +1594,252 @@ fn test_validator_saves_tower() {
     let tower4 = Tower::restore(&ledger_path, &validator_id).unwrap();
     trace!("tower4: {:?}", tower4);
     // should tower4 advance 1 slot compared to tower3????
-    assert_eq!(tower4.root(), tower3.root().map(|s| s + 1));
+    assert_eq!(tower4.root(), tower3.root() + 1);
+}
+
+fn open_blockstore(ledger_path: &Path) -> Blockstore {
+    Blockstore::open_with_access_type(ledger_path, AccessType::PrimaryOnly, None).unwrap_or_else(
+        |e| {
+            panic!("Failed to open ledger at {:?}, err: {}", ledger_path, e);
+        },
+    )
+}
+
+fn purge_slots(blockstore: &Blockstore, start_slot: Slot, slot_count: Slot) {
+    blockstore.purge_from_next_slots(start_slot, start_slot + slot_count);
+    blockstore.purge_slots(start_slot, start_slot + slot_count, PurgeType::Exact);
+}
+
+fn last_vote_in_tower(ledger_path: &Path, node_pubkey: &Pubkey) -> Option<Slot> {
+    let tower = Tower::restore(&ledger_path, &node_pubkey);
+    if let Err(tower_err) = tower {
+        if tower_err.is_file_missing() {
+            return None;
+        } else {
+            panic!("tower restore failed...: {:?}", tower_err);
+        }
+    }
+    // actually saved tower must have at least one vote.
+    let last_vote = Tower::restore(&ledger_path, &node_pubkey)
+        .unwrap()
+        .last_voted_slot()
+        .unwrap();
+    Some(last_vote)
```

Comment:
As @CriesofCarrots noticed (https://discord.com/channels/428295358100013066/560503042458517505/858178548916027402):

> Also, why does this method [2] call Tower::restore twice?

this is oversight according to this local commit (fyi, I retain all edits since i cloned the repo locally for accountability...)
The root ancesty of the redundant Tower::restore calls originated from this commit and then evolved significantly, without the bug not being noticed to this day... ;)
```
$ git show 765ec9c462dbe4e2282426f9230b941dbf5cb8f6
commit 765ec9c462dbe4e2282426f9230b941dbf5cb8f6
Author: Ryo Onodera <ryoqun@gmail.com>
Date:   Thu Oct 8 04:13:46 2020 +0900

    save

diff --git a/local-cluster/tests/local_cluster.rs b/local-cluster/tests/local_cluster.rs
index f07d021614..128e5071c4 100644
--- a/local-cluster/tests/local_cluster.rs
+++ b/local-cluster/tests/local_cluster.rs
@@ -1717,6 +1717,8 @@ fn do_test_optimistic_confirmation_violation_with_or_without_tower(with_tower: b
     let mut bad_vote_detected = false;
     let mut retry = 100;
     loop {
+        let tower = Tower::restore(&val_a_ledger_path, &val_a_7B);
+        if tower.is_err() { continue }
         let s = Tower::restore(&val_a_ledger_path, &val_a_7B).unwrap().last_voted_slot().unwrap();
         if val_c_slots.contains(&s) {
             bad_vote_detected = true;
```